stan/math/rev/mat/fun/cov_exp_quad.hpp

```diff
@@ -0,0 +1,305 @@
+#ifndef STAN_MATH_REV_MAT_FUN_COV_EXP_QUAD_HPP
+#define STAN_MATH_REV_MAT_FUN_COV_EXP_QUAD_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/rev/scal/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/fun/value_of.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/meta/length.hpp>
+#include <stan/math/prim/arr/meta/length.hpp>
+#include <stan/math/prim/scal/meta/length.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/type_traits.hpp>
+#include <vector>
+#include <cmath>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * This is a subclass of the vari class for precomputed
+     * gradients of cov_exp_quad.
+     *
+     * The class stores the double values for the distance
+     * matrix, pointers to the varis for the covariance
+     * matrix, along with a pointer to the vari for sigma,
+     * and the vari for l.
+     *
+     * @tparam T_x type of std::vector of elements
+     * @tparam T_sigma type of sigma
+     * @tparam T_l type of length scale
+     */
+    template <typename T_x, typename T_sigma, typename T_l>
+    class cov_exp_quad_vari : public vari {
+      public:
+        int size_;
+        int size_vech_;
```

Comment:
half-vectorization, the operator that takes a matrix A and returns its lower-triangular portion in a column vector by stacking the lower triangle of A column-wise. It includes the diagonal, though, and I don't here, so I'll change the name to something more descriptive.